app/src/main/java/com/capstone/event_finder/fragments/ExploreFragment.java

```diff
@@ -24,6 +62,133 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
+        query.setLimit(POST_LIMIT);
+        query.addDescendingOrder("createdAt");
+        query.findInBackground((bookmarks, e) -> {
+            if (e != null) {
+                Toast.makeText(getContext(), "Failed to find bookmarks", Toast.LENGTH_LONG).show();
+                return;
+            }
+            bookmarksList.clear();
+            bookmarksList.addAll(bookmarks);
+            queryUserBookmarkCategories(allBookmarkCategories, userInterestedCategories);
+        });
     }
 
+    private void queryUserBookmarkCategories(ArrayList<String> allBookmarkCategories, ArrayList<String> userInterestedCategories) {
+        for (int i = 0; i < bookmarksList.size(); i++) {
+            Bookmark bookmark = bookmarksList.get(i);
+            String bookmarkedEventCategory = bookmark.getEventCategory();
+            allBookmarkCategories.add(bookmarkedEventCategory);
+            bookmarkCategories.add(bookmarkedEventCategory);
+        }
+        calculatePoints(allBookmarkCategories, userInterestedCategories);
+    }
+
+    private void calculatePoints(ArrayList<String> allBookmarkCategories, ArrayList<String> userInterestedCategories) {
+        HashMap<String, Double> categoryCount = new HashMap<>();
+        double totalPoints = 0.0;
+        for (String category : bookmarkCategories) {
+            int occurrences = Collections.frequency(allBookmarkCategories, category);
+            boolean isUserInterestedCategory = userInterestedCategories.contains(category);
+            double points = 0.0;
+            if (isUserInterestedCategory) {
+                points += occurrences * POINTS_FOR_BOOKMARK_IN_CATEGORY_OF_INTEREST;
+                points += POINTS_FOR_CATEGORY_OF_INTEREST;
+            } else {
+                points += occurrences * POINTS_FOR_BOOKMARK_NOT_IN_CATEGORY_OF_INTEREST;
+            }
+            totalPoints += points;
+            categoryCount.put(category, points);
+        }
+        calculateEventsOfEachCategoryToQuery(categoryCount, totalPoints);
+    }
+
+    private void calculateEventsOfEachCategoryToQuery(HashMap<String, Double> categoryCount, Double totalPoints) {
+        recommendationList.clear();
+        for (String category : bookmarkCategories) {
+            int count = (int) (Math.ceil(10 * (categoryCount.get(category) / totalPoints)));
+            categoryCount.put(category, (double) count);
+            getAPIEvents(category, Integer.toString(count));
+        }
+    }
+
+    public void getAPIEvents(String category, String numberOfEventsToRetrieve) {
+        String eventSearchRegion = "en_US";
+        Long eventSearchRadiusFromUserInMeters = 40000L;
+        Long upcomingEventsOnly = (System.currentTimeMillis() / 1000L);
+        RetrofitClient.getInstance().getYelpAPI().getEvents(eventSearchRegion,
+                numberOfEventsToRetrieve,
+                upcomingEventsOnly,
+                eventSearchRadiusFromUserInMeters,
+                category,
```

Comment:
💡 Something to consider, if you end up working on a ranking algorithm and need to process data after fetching anyway, I think you can make this more efficient by making a single query, it looks like in the [API Docs](https://www.yelp.com/developers/documentation/v3/event_search) that you can `The category filter can be a list of comma delimited categories to get OR'd results`.

However, this would mean you have to handle the number of events in each category on the client side.

Reply:
Will definitely look into this!

Reply:
I also had an afterthought:  what happens if two of these requests come back at once.  Is `recommendationList.addAll(convertToList(result));` thread safe?  (Or would we run into the situation we discussed on Thursday about a banking application that runs on multiple threads?).

Reply:
Really good point - would love to discuss this in our next 1:1 and how to prevent this from happening!